### PR TITLE
feat(core): implement progress indicator

### LINKED ISSUE #3705: Add Progress Indicator to Library
CCv2 is in need of the progress indicator from Fiori.
https://ux.wdf.sap.corp/fiori-design-web/progress-indicator/
You can contact marie.brothankova@snk.de for more information.

### PR DESCRIPTION
**Pull request by mikerodonnell89**
closes #3705 

Implements the Progress Indicator in the core library.

Screen Shot 2021-11-22 at 12 27 20 PM
Screen Shot 2021-11-22 at 12 27 39 PM
Screen Shot 2021-11-22 at 12 27 51 PM

### CODE REVIEW

File: libs/core/src/lib/progress-indicator/progress-indicator.component.html [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+            }}</span>
+        </div>
+        <div class="fd-progress-indicator__remaining">
+            <span class="fd-progress-indicator__label" *ngIf="calculateProgressBarWidth() <= 50 && !valueText"
+                >{{ valueNow }}%</span
+            >
+            <span class="fd-progress-indicator__label" *ngIf="calculateProgressBarWidth() <= 50 && valueText">{{
+                valueText
+            }}</span>
+        </div>
```
Comment (Betrozov) [inline libs/core/src/lib/progress-indicator/progress-indicator.component.html:56]:
```suggestion
    <div class="fd-progress-indicator__container" *ngIf="calculateProgressBarWidth() as progressBarWidth">
        <div
            class="fd-progress-indicator__progress-bar"
            [ngClass]="{ 'fd-progress-indicator__progress-bar--animated': animate }"
            [ngStyle]="{ 'min-width': progressBarWidth + '%', width: progressBarWidth + '%' }"
        >
            <span class="fd-progress-indicator__label" *ngIf="progressBarWidth > 50 && !valueText"
                >{{ valueNow }}%</span
            >
            <span class="fd-progress-indicator__label" *ngIf="progressBarWidth > 50 && valueText">{{
                valueText
            }}</span>
        </div>
        <div class="fd-progress-indicator__remaining">
            <span class="fd-progress-indicator__label" *ngIf="progressBarWidth <= 50 && !valueText"
                >{{ valueNow }}%</span
            >
            <span class="fd-progress-indicator__label" *ngIf="progressBarWidth <= 50 && valueText">{{
                valueText
            }}</span>
        </div>
```

File: libs/core/src/lib/progress-indicator/progress-indicator.component.html [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+            </fd-popover-body>
+        </fd-popover>
+    </ng-container>
+</div>
+
+<ng-template #indicatorContainer>
+    <div class="fd-progress-indicator__container">
+        <div
+            class="fd-progress-indicator__progress-bar"
+            [ngClass]="{ 'fd-progress-indicator__progress-bar--animated': animate }"
```
Comment (Betrozov) [inline libs/core/src/lib/progress-indicator/progress-indicator.component.html:39]:
```suggestion
            [class.fd-progress-indicator__progress-bar--animated]="animate"
```

File: libs/core/src/lib/progress-indicator/progress-indicator.component.html [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+            [ngStyle]="{ 'min-width': calculateProgressBarWidth() + '%', width: calculateProgressBarWidth() + '%' }"
```
Comment (Betrozov) [inline libs/core/src/lib/progress-indicator/progress-indicator.component.html:40]:
```suggestion
            [style.min-width.%]="calculateProgressBarWidth()"
            [style.width.%]="calculateProgressBarWidth()"
```

File: libs/core/src/lib/progress-indicator/progress-indicator.component.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    /** Whether or not to animate changes to the progress bar's valueNow. */
+    @Input()
+    animate = false;
+
+    /** @hidden */
+    hasPopover = false;
+
+    /** @hidden */
+    @ViewChild('progressIndicator', { read: ElementRef })
+    _progressIndicator: ElementRef;
```
Comment (Betrozov) [inline libs/core/src/lib/progress-indicator/progress-indicator.component.ts:57]:
```suggestion
    private _progressIndicator: ElementRef;
```

**Review by Betrozov — CHANGES_REQUESTED** [review]

**Comment by Betrozov** [comment]
Hi @mikerodonnell89 
wonderful job
I added some comments please take a look at them

1. in case valueNow is more than valueMax or less than valueMin, we need to set valueNow as valueMax or valueMin

`valueNow > valueMax && valueNow = valueMax;`
`valueNow < valueMin && valueNow = valueMin;`
Screenshot 2021-11-24 at 14 35 23

2. I think it should be truncated or make a line break
Screenshot 2021-11-24 at 14 38 17

What do you think?

File: libs/core/src/lib/progress-indicator/progress-indicator.component.html (hunk 1 above)
Comment (dimamarksman) [reply libs/core/src/lib/progress-indicator/progress-indicator.component.html]: if `calculateProgressBarWidth() === 0` the template won't be rendered?

File: libs/core/src/lib/progress-indicator/progress-indicator.component.html [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    #progressIndicator
+    class="fd-progress-indicator"
+    tabindex="-1"
+    role="progressbar"
+    [attr.aria-valuemin]="valueMin"
+    [attr.aria-valuenow]="valueNow"
+    [attr.aria-valuemax]="valueMax"
+    [attr.aria-valuetext]="valueText"
+    [attr.aria-label]="valueText"
+    [ngClass]="state ? 'fd-progress-indicator--' + state : ''"
```
Comment (platon-rov) [inline libs/core/src/lib/progress-indicator/progress-indicator.component.html:11]: Angular recommends use `[class]` https://angular.io/guide/attribute-binding#binding-to-the-class-attribute

File: libs/core/src/lib/progress-indicator/progress-indicator.component.ts [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+    /** @hidden */
+    ngOnChanges(changes: SimpleChanges): void {
+        if (changes.valueText) {
+            this._handleTruncation();
+        }
+    }
+
+    /** @hidden */
+    ngOnInit(): void {
+        fromEvent(window, 'resize')
```
Comment (platon-rov) [inline libs/core/src/lib/progress-indicator/progress-indicator.component.ts:72]: Please use `resizeObservable` from `utils` instead.

File: libs/core/src/lib/progress-indicator/progress-indicator.component.html (hunk 5 above)
Comment (platon-rov) [reply libs/core/src/lib/progress-indicator/progress-indicator.component.html:11]: Some more details https://github.com/angular/angular/issues/40623

**Comment by mikerodonnell89** [comment]
> 



> Hi @mikerodonnell89 wonderful job I added some comments please take a look at them
> 
> 1. in case valueNow is more than valueMax or less than valueMin, we need to set valueNow as valueMax or valueMin
> 
> 
> 2. I think it should be truncated or make a line break
> What do you think?

Regarding the first issue, manipulating these values behind the scenes may have unintended consequences if the developer is unaware that this will be happening.  In my opinion the developer should be responsible for handling these edge cases.  However thanks for bringing this up because this brought an issue to my awareness where the progress indicator would not render at all when the valueMin was equal to the valueNow.

Second issue, I agree, I will add a white-space rule to this span.

**Review by dimamarksman — APPROVED** [review]

**Review by Betrozov — APPROVED** [review]

File: libs/core/src/lib/progress-indicator/progress-indicator.component.ts [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        if (changes.valueText) {
+            this._handleTruncation();
+        }
+    }
+
+    /** @hidden */
+    ngOnInit(): void {
+        resizeObservable(this._elementRef.nativeElement)
+            .pipe(debounceTime(20), takeUntil(this._onDestroy$))
+            .subscribe((_) => {
```
Comment (platon-rov) [inline libs/core/src/lib/progress-indicator/progress-indicator.component.ts:75]: `() => { ... }` will work, without `_`. Unused parameter will produce linter warning.

**Review by platon-rov — APPROVED** [review]

**Review by InnaAtanasova — APPROVED** [review]

**Pull request merged**